File: ios/Classes/MyLlamaPlugin.swift
import Flutter
import UIKit

public class MyLlamaPlugin: NSObject, FlutterPlugin {
    // 實體化我們的 Objective-C++ 橋樑
    let bridge = LlamaBridge()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "my_llama_plugin", binaryMessenger: registrar.messenger())
        let instance = MyLlamaPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if call.method == "loadModel" {
            guard let args = call.arguments as? [String: Any],
                  let path = args["path"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "缺少模型路徑", details: nil))
                return
            }

            let contextSize = args["contextSize"] as? Int ?? 2048
            let gpuLayers = args["gpuLayers"] as? Int ?? 0
            let threads = args["threads"] as? Int ?? 0

            let success = bridge.loadModel(
                atPath: path,
                contextSize: Int32(contextSize),
                gpuLayers: Int32(gpuLayers),
                threads: Int32(threads)
            )
            result(success)
        } else if call.method == "generate" {
            guard let args = call.arguments as? [String: Any],
                  let prompt = args["prompt"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "缺少 prompt", details: nil))
                return
            }

            let maxTokens = args["maxTokens"] as? Int ?? 128
            let temperature = args["temperature"] as? Double ?? 0.8
            let topK = args["topK"] as? Int ?? 40
            let topP = args["topP"] as? Double ?? 0.95

            let response = bridge.generate(
                withPrompt: prompt,
                maxTokens: Int32(maxTokens),
                temperature: Float(temperature),
                topK: Int32(topK),
                topP: Float(topP)
            )
            result(response)
        } else if call.method == "disposeModel" {
            bridge.disposeModel()
            result(true)
        } else if call.method == "getPlatformVersion" {
            result("iOS " + UIDevice.current.systemVersion)
        } else {
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/com/example/my_llama_plugin/MyLlamaPlugin.kt
package com.example.my_llama_plugin

import android.os.Handler
import android.os.Looper
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class MyLlamaPlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var channel : MethodChannel
  private val executor: ExecutorService = Executors.newSingleThreadExecutor()
  private val mainHandler = Handler(Looper.getMainLooper())

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "my_llama_plugin")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "loadModel" -> {
        val path = call.argument<String>("path") ?: ""
        val contextSize = call.argument<Int>("contextSize") ?: 2048
        val gpuLayers = call.argument<Int>("gpuLayers") ?: 0
        val threads = call.argument<Int>("threads") ?: 0
        runOnNativeThread(result) {
          MyLlamaNative.loadModel(path, contextSize, gpuLayers, threads)
        }
      }
      "generate" -> {
        val prompt = call.argument<String>("prompt") ?: ""
        val maxTokens = call.argument<Int>("maxTokens") ?: 128
        val temperature = call.argument<Double>("temperature")?.toFloat() ?: 0.8f
        val topK = call.argument<Int>("topK") ?: 40
        val topP = call.argument<Double>("topP")?.toFloat() ?: 0.95f
        runOnNativeThread(result) {
          MyLlamaNative.generate(prompt, maxTokens, temperature, topK, topP)
        }
      }
      "disposeModel" -> {
        runOnNativeThread(result) {
          MyLlamaNative.disposeModel()
          true
        }
      }
      "getPlatformVersion" -> {
        result.success("Android " + android.os.Build.VERSION.RELEASE)
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    executor.shutdown()
  }

  private fun <T> runOnNativeThread(result: Result, block: () -> T) {
    executor.execute {
      try {
        val value = block()
        mainHandler.post {
          result.success(value)
        }
      } catch (error: Throwable) {
        mainHandler.post {
          result.error("NATIVE_ERROR", error.message, null)
        }
      }
    }
  }
}
